marketplace-service/src/main/java/com/axonivy/market/service/impl/VersionServiceImpl.java

```diff
@@ -52,62 +41,69 @@
-              .sorted(new ArchivedArtifactsComparator()).toList());
-      Collections.reverse(archivedArtifacts);
-      result.put(artifact.getArtifactId(), archivedArtifacts);
-    });
-    return result;
-  }
-
   public List<MavenArtifactVersionModel> getArtifactsAndVersionToDisplay(String productId, Boolean isShowDevVersion,
       String designerVersion) {
-    List<String> versionsToDisplay = VersionUtils.getVersionsToDisplay(getPersistedVersions(productId),
+    MavenArtifactVersion cache = mavenArtifactVersionRepository.findById(productId).orElse(
+        MavenArtifactVersion.builder().productId(productId).build());
+    List<String> versionsToDisplay = VersionUtils.getVersionsToDisplay(new ArrayList<>(cache.getProductArtifactsByVersion().keySet()),
         isShowDevVersion, designerVersion);
-    List<MavenArtifact> artifactsFromMeta = getArtifactsFromMeta(productId);
-    MavenArtifact productArtifact = artifactsFromMeta.stream()
-        .filter(artifact -> artifact.getArtifactId().endsWith(MavenConstants.PRODUCT_ARTIFACT_POSTFIX)).findAny()
-        .orElse(new MavenArtifact());
-    artifactsFromMeta.remove(productArtifact);
-    Map<String, List<ArchivedArtifact>> archivedArtifactsMap = getArchivedArtifactMapFromProduct(artifactsFromMeta);
-    return handleArtifactForVersionToDisplay(versionsToDisplay, productId, artifactsFromMeta, archivedArtifactsMap);
+
+    List<Artifact> artifactsFromMeta = MavenUtils.filterNonProductArtifactFromList(getArtifactsFromMeta(productId));
+    List<MavenArtifactVersionModel> results = new ArrayList<>();
+
+    for (String mavenVersion : versionsToDisplay) {
+      List<MavenArtifactModel> artifactsByVersion = new ArrayList<>();
+      artifactsByVersion.addAll(MavenUtils.convertArtifactsToModels(artifactsFromMeta, mavenVersion));
+      artifactsByVersion.addAll(
+          cache.getProductArtifactsByVersion().computeIfAbsent(mavenVersion, k -> new ArrayList<>()));
+      artifactsByVersion.addAll(
+          cache.getAdditionalArtifactsByVersion().computeIfAbsent(mavenVersion, k -> new ArrayList<>()));
+
+      List<String> releasedVersions = productRepo.getReleasedVersionsById(productId);
+      String version = VersionUtils.getMavenVersionMatchWithTag(releasedVersions, mavenVersion);
+
+      if (StringUtils.isNotBlank(version)) {
+        productJsonRepo.findByProductIdAndVersion(productId,
+            version).stream().findAny().ifPresent(json ->
+            productRepo.findById(productId).ifPresent(product ->
+                productJsonContentService.updateProductJsonContent(json.getContent(), null, mavenVersion,
+                    version, product)
+            )
+        );
+
+        ProductModuleContent moduleContent =
+            productContentRepo.findByTagAndProductId(VersionUtils.convertVersionToTag(productId, version), productId);
+        moduleContent.setMavenVersions(Set.of(mavenVersion));
+        productContentRepo.save(moduleContent);
```

Comment:
we only change MavenVersions of ProductModuleContent then we must check does it change to not to avoid always save the same content to DB